Feature/improve rag performance

MehdinMasinovic opened the pull request:
Hey,
following improvements I incorporated:

- Instead of processing the text inside the search.py, I split it into dedicated classes (embedding.py to create the embeddings, processor.py to process the text chunks)
- Added new workflow.py that lets you create the embeddings only once and store it, instead of creating them from scratch for every user query.
- Switched to German variants of the Spacy, Embedding, and Cross-Encoder models.

Nothing in the general workflow changes, you still start app/main.py, and do a curl on the same URL PATH.

lebe1 commented:
I wanted to quickly approve this PR but so far I receive the same error calling either `/api/rag` or `/api/search`, whereas this does not occur in the current main branch for me:
```bash
 File "/home/funny/Documents/Winter24/Interdisciplinary-Project/simple-scientific-RAG/app/main.py", line 33, in handle_rag
    answer = model.rag(request.question)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/funny/Documents/Winter24/Interdisciplinary-Project/simple-scientific-RAG/app/model.py", line 12, in rag
    retrieved_text = es.search(question)
                     ^^^^^^^^^^^^^^^^^^^
  File "/home/funny/Documents/Winter24/Interdisciplinary-Project/simple-scientific-RAG/app/search.py", line 54, in search
    response = self.es.search(index='documents', body={
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/funny/.virtualenvs/RAG/lib/python3.12/site-packages/elasticsearch/_sync/client/utils.py", line 446, in wrapped
    return api(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/funny/.virtualenvs/RAG/lib/python3.12/site-packages/elasticsearch/_sync/client/__init__.py", line 4149, in search
    return self.perform_request(  # type: ignore[return-value]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/funny/.virtualenvs/RAG/lib/python3.12/site-packages/elasticsearch/_sync/client/_base.py", line 271, in perform_request
    response = self._perform_request(
               ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/funny/.virtualenvs/RAG/lib/python3.12/site-packages/elasticsearch/_sync/client/_base.py", line 352, in _perform_request
    raise HTTP_EXCEPTIONS.get(meta.status, ApiError)(
elasticsearch.BadRequestError: BadRequestError(400, 'search_phase_execution_exception', 'runtime error')
Connection to Elasticsearch closed!
```
So it seems that something in the search algorithm is not correctly running anymore. More specifically in this code part
```python
        # Perform a vector search on Elasticsearch
        response = self.es.search(index='documents', body={
            'query': {
                'script_score': {
                    'query': {
                        'match_all': {}
                    },
                    'script': {
                        'source': "cosineSimilarity(params.queryVector, 'embedding') + 1.0",  # +1.0 to make it positive
                        'params': {
                            'queryVector': query_embedding
                        }
                    }
                }
            },
            'size': top_k
        })
```
Did you run into this error before?

MehdinMasinovic commented:
@lebe1 Good that you point this out. The curl to the API returns an error, because I extracted the ES index creation from the search.py into a dedicated Python class. Previously, the index would be built every time the user asks a question. With my change this would only happen once at the beginning. 

I added one more line of code to the README to explain how one can build the index after setting up the environment. See [the latest commit](https://github.com/lebe1/simple-scientific-RAG/pull/2/commits/c54243fd1b9a5bb42db371f7a2d1ad68980888e0) for more information. Thanks for pointing this out!

lebe1 commented on app/embedding.py:
```diff
@@ -0,0 +1,66 @@
+
+        # Store chunks and embeddings
+        self.save_chunks(chunks)
+        self.save_embeddings(embeddings)
+
+
+    def create_embeddings(self, chunks, batch_size=8):
+        """Generate embeddings for each chunk using Sentence-BERT."""
+
+        # Catch edge case if there are no chunks
+        if not chunks:
+            return []
+
+        embeddings = []
+        with torch.no_grad():  # Disable gradient calculations
+            for i in range(0, len(chunks), batch_size):
+                batch_chunks = chunks[i:i + batch_size]
+                with autocast():  # Enable mixed precision
+                    batch_embeddings = self.model.encode(batch_chunks)  # Generate embeddings for the batch
+                embeddings.extend(batch_embeddings)  # Append batch embeddings to the main list
+
+                # Clear CUDA cache after each batch to free memory
+                torch.cuda.empty_cache()
+
+        return embeddings
+
+    @staticmethod
+    def save(embeddings):
+        np.save(os.path.join(FILE_PATH, '../data/embeddings.npy'), embeddings)
+
+    @staticmethod
+    def load(path= os.path.join(FILE_PATH, '../data/embeddings.npy')):
+        embeddings = np.load(path)
+        return embeddings
+
+if __name__ == "__main__":
+    # model_name='jinaai/jina-embeddings-v2-small-en'
+    model_name = 'jinaai/jina-embeddings-v2-base-de'
+    embedding = Embedding(model_name=model_name)
+    embedding.main()
```
This whole code would be unnecessary I assume, since it is only required for the method Embeddings.create_new_embeddings() highlighted below.

MehdinMasinovic replied:
Moving the `update-es-index` step from `workflow.py` to the setup_instructions sounds good to me!

Regarding the create_new_embeddings() method, it is only required once to create the embeddings for the `data/legal-basis.txt` using:

```bash
python workflow.py create-embeddings
```

It takes the `data/legal-basis.txt` as input, chunks it into 4kb parts, and computes the embeddings using `jinaai/jina-embeddings-v2-base-de`. This is also an operation I had to do only once, because the result is stored inside `data/embeddings.npy` and is simply loaded for future vector-search.

I took the same logic as for the `update-es-index` step, we only create it once. But here the difference is that the embedding model is quite huge, and it took 1) very long and 2) lots of resources to compute. Therefore I stored the results in a numpy array to later on load it easily. 

If the `data/legal-basis.txt` text changes in the future, we would need to re-run the `create-embeddings` workflow to regenerate the embeddings. 

Let me know if you have suggestions on how to do it differently, thanks again for pointing this out! 💯

lebe1 replied:
Thanks for the explanation! I added this quickly to the README, so we have this information easily accessible and do not lose it. To me, you can approve and merge this branch now as long as you don't see any improvements for the latest commit I did.

MehdinMasinovic replied:
Great, I made one minor modification in the README, I will go ahead and approve the MR! Thanks 🚀

lebe1 reviewed (commented):
Executing your command before let the system run smoothly for me. I shifted the command to the setup instructions, since the application worked again after a `docker compose down` and a `docker compose up -d` or am I missing something, so we have to execute everytime after startup?

MehdinMasinovic reviewed (commented):
Looks good to me, thanks 💯 🚀 I realize now that Pull-Request authors are not able to approve the MR. @lebe1 If you are fine with the final changes again, you could go ahead and merge 😃
